File: server/protocols/jmap-rfc-8621/src/test/scala/org/apache/james/jmap/core/KeywordsTest.scala
/****************************************************************
 * Licensed to the Apache Software Foundation (ASF) under one   *
 * or more contributor license agreements.  See the NOTICE file *
 * distributed with this work for additional information        *
 * regarding copyright ownership.  The ASF licenses this file   *
 * to you under the Apache License, Version 2.0 (the            *
 * "License"); you may not use this file except in compliance   *
 * with the License.  You may obtain a copy of the License at   *
 *                                                              *
 * http://www.apache.org/licenses/LICENSE-2.0                   *
 *                                                              *
 * Unless required by applicable law or agreed to in writing,   *
 * software distributed under the License is distributed on an  *
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
 * KIND, either express or implied.  See the License for the    *
 * specific language governing permissions and limitations      *
 * under the License.                                           *
 * **************************************************************/

package org.apache.james.jmap.core

import jakarta.mail.Flags
import jakarta.mail.Flags.Flag
import nl.jqno.equalsverifier.EqualsVerifier
import org.apache.james.jmap.mail.KeywordsFactory.{LENIENT_KEYWORDS_FACTORY, STRICT_KEYWORDS_FACTORY}
import org.apache.james.jmap.mail.{Keyword, Keywords}
import org.apache.james.mailbox.FlagsBuilder
import org.assertj.core.api.Assertions.assertThat
import org.scalatest.matchers.must.Matchers
import org.scalatest.wordspec.AnyWordSpec

object KeywordsTest {
  val ANY_KEYWORD = "AnyKeyword"
}

class KeywordsTest extends AnyWordSpec with Matchers {

  "Respect Bean Contract" in {
    EqualsVerifier.forClass(classOf[Keywords]).verify()
  }

  "fromMap should return empty when wrong keyword value" in {
    assertThat(LENIENT_KEYWORDS_FACTORY.fromMap(Map(KeywordsTest.ANY_KEYWORD -> false)).failed.get)
      .isInstanceOf(classOf[IllegalArgumentException])
  }

  "fromMap should return keywords fromMap String and boolean" in {
    val keywords = LENIENT_KEYWORDS_FACTORY.fromMap(Map(KeywordsTest.ANY_KEYWORD -> Keyword.FLAG_VALUE))
    assertThat(keywords.get.getKeywords).isEqualTo(Set(Keyword.of(KeywordsTest.ANY_KEYWORD).get))
  }

  "fromFlags should return keywords from allFlag" in {
    val keywords = LENIENT_KEYWORDS_FACTORY.fromFlags(new Flags(Flag.ANSWERED)).get
    assertThat(keywords.getKeywords).isEqualTo(Set(Keyword.ANSWERED))
  }

  "fromSet should return keywords from set of keywords" in {
    val keywords = LENIENT_KEYWORDS_FACTORY.fromSet(Set(Keyword.ANSWERED)).get
    assertThat(keywords.getKeywords).isEqualTo(Set(Keyword.ANSWERED))
  }

  "asFlags should build flags from keywords" in {
    assertThat(LENIENT_KEYWORDS_FACTORY.fromSet(Set(Keyword.ANSWERED)).get.asFlags).isEqualTo(new Flags(Flag.ANSWERED))
  }

  "asFlags with recent and deleted from should build flags from keywords and recent originFlags" in {
    val originFlags = FlagsBuilder.builder.add(Flag.RECENT, Flag.DRAFT).build
    val expectedFlags = FlagsBuilder.builder.add(Flag.ANSWERED, Flag.RECENT).build
    assertThat(LENIENT_KEYWORDS_FACTORY.fromSet(Set(Keyword.ANSWERED)).get.asFlagsWithRecentAndDeletedFrom(originFlags)).isEqualTo(expectedFlags)
  }

  "asFlags with recent and deleted from should build flags from keywords with deleted and recent originFlags" in {
    val originFlags = FlagsBuilder.builder.add(Flag.RECENT, Flag.DELETED, Flag.DRAFT).build
    val expectedFlags = FlagsBuilder.builder.add(Flag.ANSWERED, Flag.RECENT, Flag.DELETED).build
    assertThat(LENIENT_KEYWORDS_FACTORY.fromSet(Set(Keyword.ANSWERED)).get.asFlagsWithRecentAndDeletedFrom(originFlags)).isEqualTo(expectedFlags)
  }

  "asMap should return empty when empty map of string and boolean" in {
    assertThat(LENIENT_KEYWORDS_FACTORY.fromSet(Set.empty).isFailure).isFalse
  }

  "asMap should return map of string and boolean" in {
    val expectedMap = Map(Keyword.of("$Answered").get -> Keyword.FLAG_VALUE)
    assertThat(LENIENT_KEYWORDS_FACTORY.fromSet(Set(Keyword.ANSWERED)).get.asMap).isEqualTo(expectedMap)
  }

  "throwWhenUnsupportedKeyword should return failure when have unsupportedKeywords" in {
    assertThat(STRICT_KEYWORDS_FACTORY.fromSet(Set(Keyword.DRAFT, Keyword.DELETED)).isFailure).isTrue
  }

  "throwWhenUnsupportedKeyword should not throw when have draft" in {
    val keywords = STRICT_KEYWORDS_FACTORY.fromSet(Set(Keyword.ANSWERED, Keyword.DRAFT)).get
    assertThat(keywords.getKeywords).isEqualTo(Set(Keyword.ANSWERED, Keyword.DRAFT))
  }

  "filterUnsupported should filter" in {
    val keywords = LENIENT_KEYWORDS_FACTORY.fromSet(Set(Keyword.ANSWERED, Keyword.DELETED, Keyword.RECENT, Keyword.DRAFT)).get
    assertThat(keywords.getKeywords).isEqualTo(Set(Keyword.ANSWERED, Keyword.DRAFT))
  }

  "contains should return true when keywords contain keyword" in {
    val keywords = LENIENT_KEYWORDS_FACTORY.fromSet(Set(Keyword.SEEN)).get
    assertThat(keywords.contains(Keyword.SEEN)).isTrue
  }

  "contains should return false when keywords do not contain keyword" in {
    val keywords = LENIENT_KEYWORDS_FACTORY.fromSet(Set()).get
    assertThat(keywords.contains(Keyword.SEEN)).isFalse
  }

  "fromList should return keywords fromList of strings" in {
    val keywords = LENIENT_KEYWORDS_FACTORY.fromCollection(Set("$Answered", "$Flagged")).get
    assertThat(keywords.getKeywords).isEqualTo(Set(Keyword.ANSWERED, Keyword.FLAGGED))
  }

  "fromList should not throw on invalidKeyword for lenientFactory" in {
    assertThat(LENIENT_KEYWORDS_FACTORY.fromCollection(Set("in&valid")).get.keywords).isEqualTo(Keywords.DEFAULT_VALUE.keywords)
  }

  "fromMap should Not Throw On InvalidKeyword For LenientFactory" in {
    assertThat(LENIENT_KEYWORDS_FACTORY.fromMap(Map("in&valid" -> true)).get.keywords).isEqualTo(Keywords.DEFAULT_VALUE.keywords)
  }

  "fromFlags should Not Throw On InvalidKeyword For LenientFactory" in {
    assertThat(LENIENT_KEYWORDS_FACTORY.fromFlags(new Flags("in&valid")).get.keywords).isEqualTo(Keywords.DEFAULT_VALUE.keywords)
  }
}

File: server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/draft/model/KeywordsTest.java
/****************************************************************
 * Licensed to the Apache Software Foundation (ASF) under one   *
 * or more contributor license agreements.  See the NOTICE file *
 * distributed with this work for additional information        *
 * regarding copyright ownership.  The ASF licenses this file   *
 * to you under the Apache License, Version 2.0 (the            *
 * "License"); you may not use this file except in compliance   *
 * with the License.  You may obtain a copy of the License at   *
 *                                                              *
 *   http://www.apache.org/licenses/LICENSE-2.0                 *
 *                                                              *
 * Unless required by applicable law or agreed to in writing,   *
 * software distributed under the License is distributed on an  *
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
 * KIND, either express or implied.  See the License for the    *
 * specific language governing permissions and limitations      *
 * under the License.                                           *
 ****************************************************************/

package org.apache.james.jmap.draft.model;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import java.util.Map;

import jakarta.mail.Flags;
import jakarta.mail.Flags.Flag;

import org.apache.james.mailbox.FlagsBuilder;
import org.junit.Test;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.ImmutableSet;

import nl.jqno.equalsverifier.EqualsVerifier;

public class KeywordsTest {
    public static final String ANY_KEYWORD = "AnyKeyword";

    @Test
    public void shouldRespectBeanContract() {
        EqualsVerifier.forClass(Keywords.class).verify();
    }

    @Test
    public void fromMapShouldThrowWhenWrongKeywordValue() {
        assertThatThrownBy(() -> Keywords.lenientFactory()
            .fromMap(ImmutableMap.of(ANY_KEYWORD, false)))
            .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    public void fromMapShouldReturnKeywordsFromMapStringAndBoolean() {
        Keywords keywords = Keywords.lenientFactory()
            .fromMap(ImmutableMap.of(ANY_KEYWORD, Keyword.FLAG_VALUE));

        assertThat(keywords.getKeywords())
            .containsOnly(Keyword.of(ANY_KEYWORD));
    }

    @Test
    public void fromFlagsShouldReturnKeywordsFromAllFlag() {
        Keywords keywords = Keywords.lenientFactory()
            .fromFlags(new Flags(Flags.Flag.ANSWERED));

        assertThat(keywords.getKeywords())
            .containsOnly(Keyword.ANSWERED);
    }

    @Test
    public void fromSetShouldReturnKeywordsFromSetOfKeywords() {
        Keywords keywords = Keywords.lenientFactory()
            .fromSet(ImmutableSet.of(Keyword.ANSWERED));

        assertThat(keywords.getKeywords())
            .containsOnly(Keyword.ANSWERED);
    }

    @Test
    public void asFlagsShouldBuildFlagsFromKeywords() {
        assertThat(Keywords.lenientFactory()
                .fromSet(ImmutableSet.of(Keyword.ANSWERED))
                .asFlags())
            .isEqualTo(new Flags(Flags.Flag.ANSWERED));
    }

    @Test
    public void asFlagsWithRecentAndDeletedFromShouldBuildFlagsFromKeywordsAndRecentOriginFlags() {
        Flags originFlags = FlagsBuilder.builder()
            .add(Flag.RECENT, Flag.DRAFT)
            .build();

        Flags expectedFlags = FlagsBuilder.builder()
            .add(Flag.ANSWERED, Flag.RECENT)
            .build();

        assertThat(Keywords.lenientFactory()
                .fromSet(ImmutableSet.of(Keyword.ANSWERED))
                .asFlagsWithRecentAndDeletedFrom(originFlags))
            .isEqualTo(expectedFlags);
    }

    @Test
    public void asFlagsWithRecentAndDeletedFromShouldBuildFlagsFromKeywordsWithDeletedAndRecentOriginFlags() {
        Flags originFlags = FlagsBuilder.builder()
            .add(Flag.RECENT, Flag.DELETED, Flag.DRAFT)
            .build();

        Flags expectedFlags = FlagsBuilder.builder()
            .add(Flag.ANSWERED, Flag.RECENT, Flag.DELETED)
            .build();

        assertThat(Keywords.lenientFactory()
                .fromSet(ImmutableSet.of(Keyword.ANSWERED))
                .asFlagsWithRecentAndDeletedFrom(originFlags))
            .isEqualTo(expectedFlags);
    }

    @Test
    public void asMapShouldReturnEmptyWhenEmptyMapOfStringAndBoolean() {
        assertThat(Keywords.lenientFactory()
                .fromSet(ImmutableSet.of())
                .asMap())
            .isEmpty();
    }

    @Test
    public void asMapShouldReturnMapOfStringAndBoolean() {
        Map<String, Boolean> expectedMap = ImmutableMap.of("$Answered", Keyword.FLAG_VALUE);
        assertThat(Keywords.lenientFactory()
                .fromSet(ImmutableSet.of(Keyword.ANSWERED))
                .asMap())
            .isEqualTo(expectedMap);
    }

    @Test
    public void throwWhenUnsupportedKeywordShouldThrowWhenHaveUnsupportedKeywords() {
        assertThatThrownBy(() ->
            Keywords.strictFactory()
                .fromSet(ImmutableSet.of(Keyword.DRAFT, Keyword.DELETED)))
            .isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    public void throwWhenUnsupportedKeywordShouldNotThrowWhenHaveDraft() {
        Keywords keywords = Keywords.strictFactory()
            .fromSet(ImmutableSet.of(Keyword.ANSWERED, Keyword.DRAFT));

        assertThat(keywords.getKeywords())
            .containsOnly(Keyword.ANSWERED, Keyword.DRAFT);
    }

    @Test
    public void filterUnsupportedShouldFilter() {
        Keywords keywords = Keywords.lenientFactory()
            .fromSet(ImmutableSet.of(Keyword.ANSWERED, Keyword.DELETED, Keyword.RECENT, Keyword.DRAFT));

        assertThat(keywords.getKeywords())
            .containsOnly(Keyword.ANSWERED, Keyword.DRAFT);
    }

    @Test
    public void containsShouldReturnTrueWhenKeywordsContainKeyword() {
        Keywords keywords = Keywords.lenientFactory()
            .fromSet(ImmutableSet.of(Keyword.SEEN));

        assertThat(keywords.contains(Keyword.SEEN)).isTrue();
    }

    @Test
    public void containsShouldReturnFalseWhenKeywordsDoNotContainKeyword() {
        Keywords keywords = Keywords.lenientFactory()
            .fromSet(ImmutableSet.of());

        assertThat(keywords.contains(Keyword.SEEN)).isFalse();
    }

    @Test
    public void fromListShouldReturnKeywordsFromListOfStrings() {
        Keywords keywords = Keywords.lenientFactory()
            .fromCollection(ImmutableList.of("$Answered", "$Flagged"));

        assertThat(keywords.getKeywords())
            .containsOnly(Keyword.ANSWERED, Keyword.FLAGGED);
    }

    @Test
    public void fromListShouldNotThrowOnInvalidKeywordForLenientFactory() {
        assertThat(Keywords.lenientFactory()
            .fromCollection(ImmutableList.of("in&valid")))
            .isEqualTo(Keywords.DEFAULT_VALUE);
    }

    @Test
    public void fromMapShouldNotThrowOnInvalidKeywordForLenientFactory() {
        assertThat(Keywords.lenientFactory()
            .fromMap(ImmutableMap.of("in&valid", true)))
            .isEqualTo(Keywords.DEFAULT_VALUE);
    }

    @Test
    public void fromFlagsShouldNotThrowOnInvalidKeywordForLenientFactory() {
        assertThat(Keywords.lenientFactory()
            .fromFlags(new Flags("in&valid")))
            .isEqualTo(Keywords.DEFAULT_VALUE);
    }
}